// $houdini/artifacts/AllProposals.js
module.exports.name = "AllProposals";
module.exports.kind = "HoudiniQuery";
module.exports.hash = "0778d810d66eb19cf44d39b8942e3169";

module.exports.raw = `
	query AllProposals {
		proposals(first: 100) {
			id
			state
			ipfsHash
			creator
			executor {
				id
				authorized
				propositionThreshold
				votingDuration
			}
			targets
			values
			signatures
			calldatas
			withDelegatecalls
			startBlock
			endBlock
			governanceStrategy
			currentYesVote
			currentNoVote
			winner
			votes {
				id
				voter
				support
				votingPower
				timestamp
			}
			createdTimestamp
			executionTime
			initiatorQueueing
			initiatorExecution
			lastUpdateTimestamp
			lastUpdateBlock
			title
			shortDescription
			govContract
			totalPropositionSupply
			totalVotingSupply
			createdBlockNumber
			totalCurrentVoters
			aipNumber
			author
			discussions
		}
	}
`;

module.exports.rootType = "Query";

module.exports.selection = {
    "proposals": {
        "type": "Proposal",
        "keyRaw": "proposals(first: 100)",

        "fields": {
            "id": {
                "type": "ID",
                "keyRaw": "id"
            }
        }
    }
};
